src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright 2018 Inscope Metrics, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.utility;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+
+/**
+ * A regex replacement utility that can also replace tokens not found in the regex.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class RegexAndMapReplacer {
+    /**
+     * Replaces all instances of $n (where n is 0-9) with regex match groups, ${var} with regex capture group or variable (from map) 'var'.
```

Comment:
Not nuts about `$n` versus `${var}` since these are commonly the same thing. I think it would be cleaner to only support `${...}` and just have precedence rules (you already need them for named capture versus key). I'm guessing that capture groups would take precedence over dimensions (`variables`); but I am not entirely sure. If there is no clear answer or we want to provide flexibility I would recommend namespacing (e.g. `${dimension:key}` and `${capture:name}`)

Reply:
$n is already supported by the MappingSink.  It's probably the most common case that people are familiar with and I hesitate to remove it.  There are precedence rules.  And you are correct that the capture group is picked over a dimension of the same name.  Adding namespacing seems like overkill given that you can just change the name of the capture group (since the name isn't used for anything but the replacement).

Reply:
There is no need to maintain compatibility with the mapping sink. I think that there are a few other cases here which may be handy:

`${env:<name>}`
`${prop:<name>}`
`${capture:<name|number>}`
`${dimension:<name>}`

While the simplicity of precedence is appealing, I cannot come up with a consistent scheme that doesn't break down. The namespacing while somewhat more arduous on users is explicit and easy to debug in case of failure (versus precedence which can result in unexpected results).

If you feel strongly another way I'd like to hear your thoughts on this.